### DISCUSSION TITLE
Running Carla as a Systemd service

### DISCUSSION THREAD

**Discussion by kawazoe** [falkTX/Carla#1718] [Ideas]
I use EasyEffects for some microphone processing, but I find their platform and UI very limited. I really like how Carla tries to act more like a DAW in that regard and give you full control over plugins, with access to their UI, and a much better ability to manage how audio is patched through. I'm also very interested in its midi capabilities to remote control some parameters with a MIDI controller.

I know that Carla already supports running in headless mode, and with the OSC version, can connect to that instance over the network (which conveniently includes localhost).

I wish there was a one click solution to create a systemd service that would:
- setup the current project to be run in headless mode,
- with either a custom prefix or a prefix based on the project name,
- with the proper error handling to restart the process when a crash happens,
- and dependencies so that it starts after the audio sub-system is initialized (like after the pipewire/wireplumber services).

I'm currently trying to design the service file, so I might have a template for it soon here, but I would really like if this could be integrated directly into Carla. I just wish I didn't had to think about managing the file myself if I ever want to setup a second pipeline for other reasons.

**Comment by kawazoe** [comment]
After fiddling a bit around this problem, there are a lot of hurdles that makes this surprisingly difficult to do by hand.

## Preparations
Create a `services` folder inside `/usr/share/carla` and make sure it is owned by the `root` user and a group all users are in, and you have write permissions. For me, I used `realtime` since it is recommended to get pipewire working optimally. This needs to be done as root.

```shell
sudo mkdir /usr/share/carla/services
sudo chown root:realtime /usr/share/carla/services
sudo chmod -R u=rwx,g=rw,a=r /usr/share/carla/services
```

## Project
Create a Carla project. You can always come back and edit it later, but for now, we'll save the empty project in `/usr/share/carla/services/<prefix>.carxp`. We should be able to save and edit this file without any root privilege as our group has `rw` permissions. The `<prefix>` should be replaced with a simple description that you can quickly relate to the patch you are building. We will reuse it in many places. This can be "mic", "voip", or anything else that you'll recognize at a glance.

## Use as a device
If you want to use your Carla patch like any other input/output device in your computer, for instance, as a microphone input for Discord, you will need to do some additional configuration at this time. Assuming you are running pipewire, you will want to define a custom loopback device that auto-connect to your carla instance.

TODO: document this part.

## Service
Now that we have a project to run as a service. It's time to define the service in systemd.

```ini
[Unit]
Description=Carla - <prefix>
After=pipewire.service
BindsTo=pipewire.service
Wants=wireplumber.service

[Service]
Type=simple
Restart=always
ExecStart=/usr/bin/carla --no-gui --cnprefix=<prefix> /usr/share/carla/services/<prefix>.carxp

[Install]
WantedBy=graphical-session.target
```

You can save this service as `/etc/systemd/user/carla-<prefix>.service`.

Now, you can run:
```shell
systemctl daemon-reload
systemctl --user enable carla-<prefix>.service
systemctl --user start carla-<prefix>.service
```

## Conclusion
And voila! You can do this with different prefixes for as many patches you want to run.

If you ever need to edit your patch, you can use Carla OSC to remotely connect to the service instance, though you might have to specify custom ports in `ExecStart` for each instance. You can also stop the service, open the project with carla by specifying the same `cnprefix` to ensure it connects to your virtual devices like the service does, make your changes, save, quit carl, and start the service again.

I might turn this into a script if people are interested.

**Reply by Xuerian** [reply]
> TODO: document this part.

Is the relevant documentation for this somewhere? Pipewire is technically well documented, but from the perspective of someone entirely new to the system, I am finding it very difficult to build up the necessary understanding to put it together myself.

I would like to set up a service as you have here, but either way the first step is understanding how to load a project without the ability to disconnect redundant connections (which I do by hand every load at the moment, and as I understand Carla intentionally will never do on load)

**Reply by kawazoe** [reply]
@Xuerian Hey! Nice to see I'm not the only one interested in this process.

I mostly focused on the Carla and SystemD side in my post because creating virtual loopback devices in pipewire is, like you said, well documented ( https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Virtual-Devices ), while running Carla as a SystemD service is not exactly very common.

In the end, while I still use loopback devices for [Virtual Sinks](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Virtual-Devices), I didn't had enough of a need for Carla to finish the setup and make sure it is stable.

If you're interested in giving it a shot, you should have everything you need with those links to get it working. I'd also suggest [qpwgraph](https://github.com/rncbc/qpwgraph) which will help you visualize and test your configs.

Come back with your findings. I can also help you with specific questions if you have any.

**Reply by Xuerian** [reply]
Thanks for the links. On the subject of a node graph, I've been using Carla's Patchbay (I'm using the Carla flatpak, running on a ublue/oci distribution) Will one of the others, such as qpwgraph, provide something it doesn't?

**Reply by kawazoe** [reply]
They are very different. Carla's patchbay works within Carla's audio engine. It's not aware of other applications on your machine. Pipewire is the audio engine on your linux machine. That's what every application/device talk to, but it's not aware of the internal workings of those applications/devices. It's about the wires between your apps. Not inside them.

In other words, Carla's patch bay can help you route streams between various plugins in the interface, and expose some input and outputs usually bound to your audio interface's IO. That's a very old way of representing audio (and video) on computers.

In comparison, here's what a pipewire setup would look like for a pro streamer:
 - Route audio their your sound card's inputs to an AI denoiser,  
     then Carla for some compression,
     then discord
     and OBS.
 - Route the output of discord
     and the game you're playing
     to your bluetooth headphones.
 - Route the output of discord back into Carla for a different compression setup
     and back into OBS so your viewers can hear you and your friends
 - Route the audio from one of your browser tabs to OBS for your stream music
 - Route the audio from the game to OBS
     and to an app that enhance foot-step noise
     before routing it to your bluetooth headphone.
 - Route the audio from a different browser tab managing a streaming overlay
     to OBS.
     
Oh, and I did say video!
 - Route the video from your camera to a color correction program,
    before sending it to OBS
    and discord.
 - Route the output of OBS to a NDS virtual node that streams the video to the LAN
    to a whole different machine that does intensive CPU encoding
    and also routes the output to multiple different streaming applications for different providers.
    
Every app, every input/output on your devices, they all talk to pipewire, so they all show up in the graph, and qpwgraph is there to help visualize and interact with it.

**Reply by Xuerian** [reply]
Yeah, I have several virtual devices (Setup provided by Bazzite) for music, games, browser, etc. I run each through a dynamic range compressor as well as sidechaining the media against voice chat. It's fantastic.

And before this I used KXAudio drivers on windows with my soundblaster cards to do very similar things, and is why I'm using Carla in the first place.

But I believe you may be mistaken with Carla's Patchbay.. It provides the same view and control over the pipewire graph as qpwgraph - but also allows you to run and connect plugins hosted by Rack. I have them side by side right now, it's effectively the same functionality aside from that, as far as I can tell.

**Reply by kawazoe** [reply]
It depends on what device you setup Carla to use. As far as I know, Carla isn't a Pipewire application, so it depends on compatibility layer like PipeWirePulse or PipeWireJack to function. It will only see the inputs that are available to it's layer.

In my example, you don't need to setup specific virtual devices for music/games/browser/etc for them to show up in your graph. They are just there from the start, because every process is already it's own node.

Basically, it sounds to me like you're just reinventing what pipewire already does for you to get it to play well with Clara's patch bay, instead of just using pipewire how it's been designed to be used in the first place.

**Reply by Xuerian** [reply]
I think you should check out the patch bay and see what it can do, and compare it side-by-side to qpw.

But either way, the advice you gave is useful and hopefully I'll get to trying it out soon

**Reply by kawazoe** [reply]
I'm aware of what the Carla patchbay can do. It can't hook into other applications directly; that's just the nature of ALSA, the legacy sound pipeline on linux. Carla supports JACK, so other applications that supports JACK can communicate directly, but that's usually professional audio applications and not stuff like Chrome or MPV. Unless you take the time to explicitely setup custom virtual devices for every single one of your applications, like you mention you did earlier, Carla have no way to know about them. If you look at the patchbay screenshot from the readme, you'll see what I mean. You can see "system" nodes, and "jack" nodes, but you don't get one node per instance of Chrome playing audio on your system. You get the *devices* that are available on your system, not individual *processes*.

**Reply by falkTX** [reply]
if you use pipewire and carla in jack mode, everything can talk with everything else. I can have firefox sending audio into carla and vice-versa, because pipewire client streams can be connected at will, no matter if they use pulse, jack or native pipewire apis

**Reply by Xuerian** [reply]
![20240202_163241_Untitled1 — A PipeWire Graph Qt GUI Interface](https://github.com/falkTX/Carla/assets/2985/a23e0d5d-f2f0-4f34-bce8-2ac9ad962289)
![20240202_163304_Carla - test carxp](https://github.com/falkTX/Carla/assets/2985/d8a4d4a3-aa63-41de-8ff9-1b8c553c95ef)

For my use case, the Patchbay seemingly does most of what you describe, and also lets me configure plugins.

I don't mean to dismiss qpwgraph by the comparison.

This setup works pretty well for me and is quite similar (albeit somewhat simplified) from what I used in windows with the aging KXAudio drivers and a soundblaster.

**Reply by kawazoe** [reply]
@falkTX I guess you would need the pipewire-jack compatibility package for this to work, right? If that's the case then, there are a lot of advantages on using pipewire directly instead of the jack compatibility layer. Namely, pipewire's graph has authorization built-in, which might or might not matter to you. You also save on buffer copies as pipewire need to copy the stream buffer to support jack. Else, then I'm missing something. Pipewire streams cannot connect to other applications unless they are using a compatible API. This means that unless pipewire support has been added since I opened this thread, you can't connect to a pure pipewire stream from Carla without either pipewire-pulse or pipewire-jack installed. But then, you're not doing pipewire. You're doing pulse or jack.

**Reply by falkTX** [reply]
you need pipewire-jack indeed yes. and there is no point on trying to use pipewire apis directly since this jack mode works very well, and also allows people that do not use pipewire to keep using it in good old jack mode.
I have no intentions on doing a pipewire-specific backend, what we have through jack works well and is more compatible.

PS: the app name is carla not clara.

**Reply by kawazoe** [reply]
Oh yes of course. I don't think you should spend time explicitely supporting pipewire. I was just trying to explain how the Carla patchbay is different to pipewire, and what you just said confirmed my suspitions.

With that being said, I don't think that "jack works well and is more compatible" is a good argument to avoid supporting it. The pipewire FAQ goes very deep on [why pipewire can be better than jack](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#how-is-pipewire-supposed-to-be-a-better-jack) for some usecases. I think the biggest one for Carla relates to performance. It's faster than jack with similar latencies, which is good for audio production because it means less overhead, and it also supports dynamic latencies which is good for power constrained devices like laptops. They *can* be good reasons to want to support pipewire direcly, but only if your users needs those things. Until you get tons of people trying out setups like me and Xueran, who might benefit from running in dynamic latency mode, jack's definitely the way to go.

Nice chat though :heart: 

PS: Oh wow, I didn't even noticed that I switched midway though the conversation. Fixed :)

**Reply by falkTX** [reply]
all is fine and good except

> It's faster than jack with similar latencies, which is good for audio production because it means less overhead

is simply not true. pipewire has a lot more stuff to deal with, so "less overhead" is just not possible, and "faster" in this context I dont know what you mean.

**Reply by kawazoe** [reply]
I'm really just quoting their wiki here. They have benchmarks that demonstrates pipewire needs about half the CPU instructions to achieve the same kind of work than JACK would for a simple usecase like playback/recording without any kind of resampling.

But even if that would be wrong, it doesn't change that pipewire is a bit like the Wayland of audio. Every distro out there is slowly switching to pipewire. Some tools like Hyprland flatout requires it if you want to do screensharing. Over time, it's going to be harder and harder to find a system that won't already be runing pipewire, which means the JACK implementation you're connecting to will most likely be a compatibility layer on top of a pipewire node, like in the scenario we're discussing here. In the end, you'll have to pay the pipewire tax anyway so a JACK Carla audio system will always be less efficient than a pure pipewire one.

But again, like I said, it probably doesn't matter much, and for at least a decade people who want a pure JACK setup to run their DAW with the best performance will still be able to make a system without pipewire if they really wanted to.
